src/rgw/rgw_string.h

```diff
@@ -233,4 +238,167 @@ extern bool match_wildcards(boost::string_view pattern,
                             boost::string_view input,
                             uint32_t flags = 0);
 
+inline void rgw_escape_str(const std::string& s, const char esc_char, const char special_char, std::string& dest)
+{
+ dest.reserve(s.size() * 2);
+ 
+ auto dest_iter = begin(dest);
+ 
+ for (const char c : s) {
+    if(esc_char == c || special_char == c) {
+        *dest_iter++ = esc_char;
+    }
+    
+    *dest_iter++ = c;
+ }
+}
+
+inline void rgw_escape_str(const std::string& s, const char esc_char, const char special_char, std::string *dest_ptr)
+{
+ return rgw_escape_str(s, esc_char, special_char, *dest_ptr);
+}
+
+inline std::string rgw_escape_str(const std::string& s, const char esc_char, const char special_char)
+{
+ std::string result;
+ rgw_escape_str(s, esc_char, special_char, result);
+ return result;
+}
+
+inline ssize_t rgw_unescape_str(const std::string& src, const ssize_t ofs,
+                       const char esc_char, const char special_char,
+                       std::string& dest)
+{
+  dest.reserve(src.size());
+  auto destp = begin(dest);
+
+  bool esc = false;
+
+  for (size_t i = ofs; i < src.size(); i++) {
+    char c = src[i];
+    if (!esc && c == esc_char) {
+      esc = true;
+      continue;
+    }
+    if (!esc && c == special_char) {
+      dest.resize(i);
+      return dest.size();
+    }
+    *destp++ = c;
+    esc = false;
+  }
+
+  return std::string::npos;
+}
+
+inline ssize_t rgw_unescape_str(const std::string& s, const ssize_t ofs,
+                       const char esc_char, const char special_char,
+                       std::string *dest_ptr)
+{
+  return rgw_unescape_str(s, ofs, esc_char, special_char, *dest_ptr);
+}
+
+class CephContext;
+
+/* size should be the required string size + 1 */
+int gen_rand_base64(CephContext *cct, char *dest, int size);
+
+void gen_rand_alphanumeric_lower_CHAR(CephContext *cct, char *dest, const int nchars);
+
+namespace detail {
+
+// Does NOT null-terminate:
+template <typename OutputIterT, typename SizeT>
+void gen_rand_buffer_mutate_in_place(OutputIterT& oi, const SizeT size, std::string_view tbl)
+{
+  std::generate_n(oi, size, [&]() noexcept {
+                  auto n = ceph::util::generate_random_number(0, tbl.size());
+                  return tbl[n];
+                 });
+}
+
+inline std::string gen_rand_string(const size_t nchars, std::string_view tbl)
+{
+ std::string result;
+
+ result.resize(nchars);
+
+ return std::generate_n(back_inserter(result), result.size(), [&]() noexcept {
+                        auto n = ceph::util::generate_random_number(0, tbl.size());
```

Comment:
please wrap at 80 chars.

Reply:
This code will go away in the next changeset.